bin/cron/pardot_import

```diff
@@ -0,0 +1,212 @@
+  # query parameter)
+  url = build_create_batch_prospects_url(prospects)
+
+  # Post the data
+  post_with_auth_retry(url)
+
+  # build array of the emails we just created in Pardot
+  prospect_emails = prospects.collect{ |x| x["email"] }.to_a
+  # store them in our DB as created in Pardot so we don't double-create them.
+  PEGASUS_DB[:contact_rollups].where("email in ?", prospect_emails).update("pardot_sync_at=NOW()")
+end
+
+# Build the URL to batch create prospects in Pardot. The prospect data is a JSON blob in the
+# query param.
+def build_create_batch_prospects_url(prospects)
+  prospects_payload = { prospects: prospects }
+  prospects_payload_json = prospects_payload.to_json
+  prospects_payload_json_encoded = URI.encode(prospects_payload_json)
+  # We also need to encode plus signs in email addresses, otherwise Pardot rejects them as invalid.
+  # URI.encode does not encode plus signs, as they are valid characters in the base of a URL.
+  # (Although they are NOT valid in the query string, which is where this data is going.)
+  prospects_payload_json_encoded = prospects_payload_json_encoded.gsub("+", "%2B")
+
+  "#{PARDOT_BATCH_CREATE_URL}?prospects=#{prospects_payload_json_encoded}"
+end
+
+def post_with_auth_retry(url)
+  # do the post to Pardot
+  post_request(url)
+rescue InvalidApiKeyException
+  # If we fail with an invalid key, that probably means our API key has expired. Try one time
+  # to request a new API key and try the post again. If that fails, that is a fatal error.
+  request_pardot_api_key
+  post_request(url)
+end
+
+# Make an API request
+def post_request(url)
+  request_pardot_api_key if $api_key.nil?
+  # add the API key and user key parameters to the URL
+  auth_url = append_auth_params_to_url(url)
+  uri = URI(auth_url)
+  response = Net::HTTP.post_form(uri, {})
+
+  puts "Response: " + response.to_s
+
+  xml = Nokogiri::XML(response.body, &:noblanks)
+  raise "Pardot create user response did not return parsable XML" if xml.nil?
+
+  error_details = xml.xpath('/rsp/err').text
+  raise InvalidApiKeyException if error_details.include? "Invalid API key or user key"
+
+  status = xml.xpath('/rsp/@stat').text
+  raise "Pardot create user response did not include status" if status.nil?
+
+  raise "Pardot create user response failed with status " + status + " " + xml.to_s if status != "ok"
+end
+
+def append_auth_params_to_url(url)
+  "#{url}&api_key=#{$api_key}&user_key=#{CDO.pardot_user_key}"
+end
+
+def main
+  return unless only_one_running?(__FILE__)
+  prospects = []
+
+  # query the conact rolliups
+  PEGASUS_DB[:contact_rollups].where("roles like '%Teacher%' AND pardot_sync_at IS NULL").each do |contact_rollup|
+    puts contact_rollup[:email]
+
+    # Pardot explicitly considers invalid any email addresses that end with a period in the base
+    # (before the @). We have at least one of these, which appears to be a typo.
+    next if contact_rollup[:email].include? ".@"
```

Comment:
Note we have a number of invalid email addresses (for this and other reasons) in the pegasus `contacts` table (likely typos). We never remove entries from that table, even when failing to send, and we have fairly poor email format validation in the various sources (forms, code studio account page, workshop signup, etc) at the moment.